### ISSUE TITLE
P4-C: Skill prompt optimizer using rubric reward signals

### ISSUE THREAD

**Issue by jpequegn** [jpequegn/rubric-gates#29]
## Description — Phase C: Skill Prompt Optimization

**Goal:** Use accumulated rubric scores as reward signals to automatically optimize your Claude Code skill prompts, so skills produce higher-quality code over time.

**The RLVR loop here:** skill prompt generates code → rubric scores it → scores guide prompt mutation → better prompts → better code.

## Why This Matters

This is the highest-leverage application for your team. Instead of training a new model (Phase B), you optimize the *prompts* that drive Claude — which is the model you're already using. Benefits:
- **No GPU needed for inference** — optimized prompts run on Claude as before
- **Direct improvement** to your existing skill ecosystem
- **Measurable ROI** — track score improvement per skill over time
- **Automatic** — skills get better without manual prompt tweaking

### DSPy-style Prompt Optimization

Treat each skill prompt as a program with tunable components. Use rubric scores as the objective function to optimize those components.

```
1. Collect: score history per skill from P1 storage
2. Identify: which skills have the lowest average scores
3. Mutate: generate prompt variations for the weakest skills
4. Evaluate: run each variation on test tasks, score with rubric engine
5. Select: keep the variation with the highest rubric scores
6. Deploy: update the skill with the optimized prompt
7. Monitor: track score improvement via P1 dashboard
```

### Prompt Mutation Strategies

| Strategy | How It Works | When to Use |
|----------|-------------|-------------|
| **Instruction refinement** | Add/modify instructions about quality, security, structure | Low maintainability or documentation scores |
| **Example injection** | Add few-shot examples of high-scoring code | Model needs concrete patterns to follow |
| **Constraint addition** | Add explicit constraints ("never hardcode credentials") | Security or specific pattern violations |
| **Output format tuning** | Adjust requested output format (comments, structure) | Format-related low scores |
| **Chain-of-thought** | Add reasoning prompts ("think about security before writing") | Broad quality improvement |

### Optimization Algorithm

```python
def optimize_skill(skill_name: str, test_tasks: list[str]) -> OptimizedSkill:
    """Optimize a skill prompt using rubric-guided search."""
    
    current_prompt = load_skill_prompt(skill_name)
    current_scores = evaluate_skill(current_prompt, test_tasks)
    
    for iteration in range(max_iterations):
        # Generate mutations
        candidates = generate_mutations(current_prompt, current_scores)
        
        # Evaluate each candidate
        for candidate in candidates:
            scores = evaluate_skill(candidate.prompt, test_tasks)
            candidate.avg_score = mean(scores)
        
        # Select best
        best = max(candidates, key=lambda c: c.avg_score)
        
        if best.avg_score > current_scores.mean() + min_improvement:
            current_prompt = best.prompt
            current_scores = best.scores
            log_improvement(skill_name, iteration, best)
        else:
            break  # Converged
    
    return OptimizedSkill(prompt=current_prompt, scores=current_scores)
```

## Interface

```python
class SkillOptimizer:
    """Optimize Claude Code skill prompts using rubric rewards."""
    
    def __init__(self, rubric_engine: RubricEngine, config: OptimizerConfig):
        ...
    
    def analyze_skill(self, skill_name: str) -> SkillAnalysis:
        """Analyze a skill's scoring history, identify weakest dimensions."""
    
    def generate_mutations(self, prompt: str, analysis: SkillAnalysis) -> list[PromptCandidate]:
        """Generate prompt variations targeting weak dimensions."""
    
    def evaluate_candidate(self, candidate: PromptCandidate, test_tasks: list[str]) -> EvalResult:
        """Run candidate prompt on test tasks, score with rubric."""
    
    def optimize(self, skill_name: str, test_tasks: list[str]) -> OptimizedSkill:
        """Run full optimization loop for a skill."""
    
    def compare(self, original: str, optimized: str, test_tasks: list[str]) -> ComparisonReport:
        """Side-by-side comparison of original vs optimized on rubric scores."""

class SkillOptimizationRunner:
    """Batch optimizer for the entire skill ecosystem."""
    
    def identify_candidates(self, min_score_history: int = 20) -> list[str]:
        """Find skills with enough data and low scores — best candidates for optimization."""
    
    def run_batch(self, skill_names: list[str]) -> BatchReport:
        """Optimize multiple skills, generate improvement report."""
    
    def deploy(self, skill_name: str, optimized: OptimizedSkill, backup: bool = True) -> None:
        """Deploy optimized prompt to the live skill (with backup)."""
```

## Safety Rails

Prompt optimization can go wrong. Safeguards:

- **A/B testing required:** Optimized prompts run alongside originals (P3's A/B framework)
- **Regression detection:** If any dimension drops > 0.1 from baseline, reject the mutation
- **Human approval gate:** Optimized prompts are *suggested*, not auto-deployed by default
- **Rollback:** Every deployed optimization has a one-command rollback
- **Diversity constraint:** Optimized prompts must still produce varied outputs (not overfit to test tasks)

## CLI Commands

```bash
# Analyze which skills would benefit most from optimization
python -m training.skill_optimizer analyze

# Optimize a specific skill
python -m training.skill_optimizer optimize --skill scaffold-api --iterations 10

# Compare original vs optimized
python -m training.skill_optimizer compare --skill scaffold-api

# Deploy (with backup)
python -m training.skill_optimizer deploy --skill scaffold-api

# Rollback
python -m training.skill_optimizer rollback --skill scaffold-api
```

## Expected Outcomes

| Metric | Before Optimization | After (target) |
|--------|-------------------|-----------------|
| Avg composite score per skill | Varies (0.5-0.7) | +0.1 improvement |
| Skills in green tier by default | ~50% | > 75% |
| Security findings per generation | ~0.3 | < 0.1 |
| Manual prompt tuning effort | Hours per skill | Automated |

## Acceptance Criteria

- [ ] Optimizer in `training/skill_optimizer.py`
- [ ] Runner in `training/skill_runner.py`
- [ ] At least 3 mutation strategies implemented
- [ ] Optimization loop converges (measurable improvement on test tasks)
- [ ] Safety rails: regression detection, human approval gate, rollback
- [ ] CLI for analyze/optimize/compare/deploy/rollback
- [ ] Integration with P1 storage (reads score history)
- [ ] Comparison report shows before/after per dimension
- [ ] Tests with mock skill prompts and rubric engine

## Dependencies

- #26 (training infrastructure)
- #9 (rubric engine — evaluates prompt candidates)
- #3 (storage — reads skill score history)
- #11 (dashboard — reports on optimization results)
- #27 (Phase A — optionally use distilled scorer for faster evaluation during search)
